### PR TITLE
Nacos客户端磁盘降级需求  issues:11053

### PR DESCRIPTION
**Pull request by guozongkang**
Please do not create a Pull Request without creating an issue first.

## What is the purpose of the change

https://github.com/alibaba/nacos/issues/11053

## Brief changelog

1.使用磁盘来实现Nacos客户端降级方案
2.可以修改机器上的文件，开启或关闭降级，并且指定需要降级的服务
3.查询机器列表时，优先走降级逻辑

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

* [ ] Make sure there is a Github issue filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
* [ ] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
* [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
* [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/alibaba/nacos/tree/master/test).
* [ ] Run `mvn -B clean package apache-rat:check findbugs:findbugs -Dmaven.test.skip=true` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.

### CODE REVIEW

File: client/src/main/java/com/alibaba/nacos/client/naming/backups/FailoverReactor.java [hunk 1]
```diff
@@ -107,184 +101,121 @@ public void init() {
...
-            try {
-                File switchFile = Paths.get(failoverDir, UtilAndComs.FAILOVER_SWITCH).toFile();
-                if (!switchFile.exists()) {
-                    switchParams.put(FAILOVER_MODE_PARAM, Boolean.FALSE.toString());
-                    NAMING_LOGGER.debug("failover switch is not found, {}", switchFile.getName());
-                    return;
+
+            FailoverSwitch fSwitch = failoverDataSource.getSwitch();
+            if (fSwitch != null && fSwitch.getEnabled() && !failoverSwitchEnable) {
+                failoverSwitchEnable = Boolean.TRUE;
```
Comment (nkorange) [inline client/src/main/java/com/alibaba/nacos/client/naming/backups/FailoverReactor.java:121]: 直接failoverSwitchEnable=true即可

File: client/src/main/java/com/alibaba/nacos/client/naming/backups/FailoverReactor.java [hunk 2]
```diff
@@ -107,184 +101,121 @@ public void init() {
...
-                }
-                
-                File[] files = cacheDir.listFiles();
-                if (files == null) {
-                    return;
+
+            if (fSwitch != null && failoverSwitchEnable && !fSwitch.getEnabled()) {
+                failoverSwitchEnable = Boolean.FALSE;
+                Map<String, ServiceInfo> map = new ConcurrentHashMap<>(200);
+                Map<String, FailoverData> failoverData = failoverDataSource.getFailoverData();
```
Comment (nkorange) [inline client/src/main/java/com/alibaba/nacos/client/naming/backups/FailoverReactor.java:138]: 这里似乎只需要将serviceMap和serviceInfoHolder.getServiceInfoMap()就行了？

File: client/src/main/java/com/alibaba/nacos/client/naming/backups/FailoverReactor.java [hunk 3]
```diff
@@ -107,184 +101,121 @@ public void init() {
...
             }
+
+            serviceMap.clear();
         }
     }
-    
-    class DiskFileWriter extends TimerTask {
-        
+
+    class FailoverDataWriter extends TimerTask {
```
Comment (nkorange) [inline client/src/main/java/com/alibaba/nacos/client/naming/backups/FailoverReactor.java:159]: 根据 @KomachiSion 的建议，把这个writer任务去掉

File: client/src/main/java/com/alibaba/nacos/client/naming/backups/FailoverDataSource.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+     * @return map
+     */
+    Map<String, FailoverData> getFailoverData();
+
+    /**
+     * Save disaster recovery data.
+     *
+     * @param failoverDataMap
+     */
+    void saveFailoverData(Map<String, FailoverData> failoverDataMap);
```
Comment (nkorange) [inline client/src/main/java/com/alibaba/nacos/client/naming/backups/FailoverDataSource.java:47]: 根据 @KomachiSion 的建议，把这个方法去掉

File: client/src/main/java/com/alibaba/nacos/client/naming/NacosNamingService.java [hunk 5]
```diff
@@ -183,119 +178,137 @@ public void deregisterInstance(String serviceName, String groupName, String ip,
...
-            boolean subscribe) throws NacosException {
+                                          boolean subscribe) throws NacosException {
         ServiceInfo serviceInfo;
+        List<Instance> list;
         String clusterString = StringUtils.join(clusters, ",");
-        if (subscribe) {
-            serviceInfo = serviceInfoHolder.getServiceInfo(serviceName, groupName, clusterString);
-            if (null == serviceInfo || !clientProxy.isSubscribed(serviceName, groupName, clusterString)) {
-                serviceInfo = clientProxy.subscribe(serviceName, groupName, clusterString);
+        if (serviceInfoHolder.isFailoverSwitch()) {
```
Comment (nkorange) [inline client/src/main/java/com/alibaba/nacos/client/naming/NacosNamingService.java:236]: 除了getAllInstances，其他的查询实例的方法例如selectInstances也需要保护容灾逻辑

Comment (nkorange) [inline client/src/main/java/com/alibaba/nacos/client/naming/NacosNamingService.java:236]: 这里还要判断当前查询的服务名是否在容灾开关里

File: client/src/main/java/com/alibaba/nacos/client/naming/NacosNamingService.java [hunk 6]
```diff
@@ -183,119 +178,137 @@ public void deregisterInstance(String serviceName, String groupName, String ip,
...
+            list = new ArrayList<>();
+            serviceInfo = serviceInfoHolder.getFailoverServiceInfo(serviceName, groupName, clusterString);
+            if (serviceInfo != null && serviceInfo.getHosts().size() > 0) {
+                serviceInfo.getHosts();
```
Comment (nkorange) [inline client/src/main/java/com/alibaba/nacos/client/naming/NacosNamingService.java:240]: 这一行啥意思？

File: client/src/main/java/com/alibaba/nacos/client/naming/NacosNamingService.java [hunk 7]
```diff
@@ -183,119 +178,137 @@ public void deregisterInstance(String serviceName, String groupName, String ip,
...
+                if (!clientProxy.isSubscribed(serviceName, groupName, clusterString)){
+                    serviceInfo = clientProxy.subscribe(serviceName, groupName, clusterString);
+                }
+            } else{
+                if (subscribe) {
+                    serviceInfo = serviceInfoHolder.getServiceInfo(serviceName, groupName, clusterString);
```
Comment (nkorange) [inline client/src/main/java/com/alibaba/nacos/client/naming/NacosNamingService.java:246]: serviceInfoHolder.getServiceInfo方法里也有判断容灾的逻辑，是否应该移除？

File: client/pom.xml [hunk 8]
```diff
@@ -119,6 +119,10 @@
             <groupId>org.yaml</groupId>
             <artifactId>snakeyaml</artifactId>
         </dependency>
+        <dependency>
```
Comment (nkorange) [inline client/pom.xml:122]: 运行mvn -Prelease-nacos -Dmaven.test.skip=true clean install -U 不成功

File: client/src/main/java/com/alibaba/nacos/client/naming/NacosNamingService.java [hunk 9]
```diff
@@ -183,119 +178,137 @@ public void deregisterInstance(String serviceName, String groupName, String ip,
...
+                    if (null == serviceInfo || !clientProxy.isSubscribed(serviceName, groupName, clusterString)) {
+                        serviceInfo = clientProxy.subscribe(serviceName, groupName, clusterString);
+                    }
+                }
```
Comment (nkorange) [inline client/src/main/java/com/alibaba/nacos/client/naming/NacosNamingService.java:250]: 这里少了subscribe=false时的处理

File: client/src/main/java/com/alibaba/nacos/client/naming/backups/FailoverReactor.java [hunk 10]
```diff
@@ -86,16 +80,16 @@ public FailoverReactor(ServiceInfoHolder serviceInfoHolder, String cacheDir) {
...
-        
-        executorService.scheduleWithFixedDelay(new DiskFileWriter(), 30, DAY_PERIOD_MINUTES, TimeUnit.MINUTES);
-        
+
+        executorService.scheduleWithFixedDelay(new FailoverSwitchRefresher(), 0L, 5000L, TimeUnit.MILLISECONDS);
+
+        executorService.scheduleWithFixedDelay(new FailoverDataWriter(), 30, DAY_PERIOD_MINUTES, TimeUnit.MINUTES);
+
         // backup file on startup if failover directory is empty.
         executorService.schedule(() -> {
```
Comment (nkorange) [inline client/src/main/java/com/alibaba/nacos/client/naming/backups/FailoverReactor.java:94]: 这个任务也去掉吧

File: client/src/main/java/com/alibaba/nacos/client/naming/backups/FailoverReactor.java [hunk 11]
```diff
@@ -53,30 +43,34 @@
...
+
+    private final MultiGauge failoverInstanceCounts;
+
+    private String notifierEventScope;
+
+    public FailoverReactor(ServiceInfoHolder serviceInfoHolder, String cacheDir, String notifierEventScope) {
         this.serviceInfoHolder = serviceInfoHolder;
         this.failoverDir = cacheDir + FAILOVER_DIR;
+        this.notifierEventScope = notifierEventScope;
+        failoverDataSource = new DiskFailoverDataSource(serviceInfoHolder, cacheDir);
```
Comment (nkorange) [inline client/src/main/java/com/alibaba/nacos/client/naming/backups/FailoverReactor.java:72]: 这里需要通过反射来获取当前实现类

File: client/src/main/java/com/alibaba/nacos/client/naming/backups/FailoverReactor.java [hunk 12]
```diff
@@ -53,30 +43,34 @@
...
+
+    private final FailoverDataSource failoverDataSource;
+
+    private final MultiGauge failoverInstanceCounts;
+
+    private String notifierEventScope;
+
+    public FailoverReactor(ServiceInfoHolder serviceInfoHolder, String cacheDir, String notifierEventScope) {
         this.serviceInfoHolder = serviceInfoHolder;
         this.failoverDir = cacheDir + FAILOVER_DIR;
```
Comment (nkorange) [inline client/src/main/java/com/alibaba/nacos/client/naming/backups/FailoverReactor.java:70]: failoverDir不是FailoverReactor的必须字段了

**Review by nkorange — CHANGES_REQUESTED** [review]

File: client/src/main/java/com/alibaba/nacos/client/naming/NacosNamingService.java (hunk 6 above)
Comment (guozongkang) [reply client/src/main/java/com/alibaba/nacos/client/naming/NacosNamingService.java:243]: 判断，当前服务的服务列表是否在容灾开关里

File: client/src/main/java/com/alibaba/nacos/client/naming/NacosNamingService.java
Comment (KomachiSion) [inline client/src/main/java/com/alibaba/nacos/client/naming/NacosNamingService.java:1]: 大量无效改动，请使用nacos code style进行reformat

File: client/src/main/java/com/alibaba/nacos/client/naming/cache/ServiceInfoHolder.java
Comment (KomachiSion) [inline client/src/main/java/com/alibaba/nacos/client/naming/cache/ServiceInfoHolder.java:1]: 同 NacosNamingService.java

**Review by KomachiSion — CHANGES_REQUESTED** [review]

**Review by nkorange — DISMISSED** [review]

File: client/src/main/java/com/alibaba/nacos/client/naming/NacosNamingService.java [hunk 13]
```diff
@@ -87,7 +86,6 @@ public NacosNamingService(Properties properties) throws NacosException {
     }
     
     private void init(Properties properties) throws NacosException {
-        PreInitUtils.asyncPreLoadCostComponent();
```
Comment (KomachiSion) [inline client/src/main/java/com/alibaba/nacos/client/naming/NacosNamingService.java:90]: 为什么删除这个？

File: client/src/main/java/com/alibaba/nacos/client/naming/NacosNamingService.java [hunk 14]
```diff
@@ -326,6 +305,30 @@ private List<Instance> selectInstances(ServiceInfo serviceInfo, boolean healthy)
         return list;
     }
     
+    private ServiceInfo getServiceInfo(String serviceName, String groupName, List<String> clusters, boolean subscribe)
```
Comment (KomachiSion) [inline client/src/main/java/com/alibaba/nacos/client/naming/NacosNamingService.java:308]: 能否再优化一下这段逻辑， 看着乱，实在不行把从failover获取的逻辑和普通获取的逻辑分到两个方法里。

File: client/src/main/java/com/alibaba/nacos/client/naming/backups/FailoverReactor.java [hunk 15]
```diff
@@ -92,189 +87,61 @@ public FailoverReactor(ServiceInfoHolder serviceInfoHolder, String cacheDir) {
...
-                        .equals(serviceInfo.getName(), UtilAndComs.ALL_HOSTS)) {
-                    continue;
+            if (fSwitch != null && failoverSwitchEnable && !fSwitch.getEnabled()) {
+                failoverSwitchEnable = false;
+                Map<String, ServiceInfo> serviceInfoMap = serviceInfoHolder.getServiceInfoMap();
+                for (Map.Entry<String, ServiceInfo> entry : serviceMap.entrySet()) {
+                    ServiceInfo oldService = entry.getValue();
+                    ServiceInfo newService = serviceInfoMap.get(entry.getKey());
+                    boolean changed = serviceInfoHolder.isChangedServiceInfo(oldService, newService);
+                    if (changed) {
```
Comment (KomachiSion) [inline client/src/main/java/com/alibaba/nacos/client/naming/backups/FailoverReactor.java:132]: FailOver加了通知机制， 是否会导致在和服务端来回抢通知？

另外，当failover被关闭的时候，是否要触发一次通知，让订阅者感知到服务端推送的数据。

**Review by KomachiSion — CHANGES_REQUESTED** [review]

File: client/src/main/java/com/alibaba/nacos/client/naming/backups/FailoverReactor.java (hunk 15 above)
Comment (nkorange) [reply client/src/main/java/com/alibaba/nacos/client/naming/backups/FailoverReactor.java:132]: @KomachiSion 这个地方就是当failover关闭时触发通知，你看125行的if语句

**Review by KomachiSion — APPROVED** [review]
Will be merged after 2.3.0 released.

**Pull request merged**